File: fsharp/list-ops/ListOps.fs
module ListOps

let rec foldl fn init list =
    match list with
    | [] -> init
    | x::xs -> foldl fn (fn init x) xs
let rec foldr fn init list =
    list
    |> List.rev
    |> foldl (fun acc n -> fn n acc) init
    
let length list = foldl (fun acc _ -> acc + 1) 0 list
let reverse list = foldl (fun acc n -> n :: acc) [] list
let map f list = foldr (fun n acc -> f n :: acc) [] list
let filter f list = foldr (fun n acc -> if f n then n :: acc else acc) [] list
let append xs ys = foldr (fun n acc -> n :: acc) ys xs
let concat xs = foldr append [] xs

File: dart/list-ops/lib/list_ops.dart
extension ListOps on List {
  void append<T>(List<T> other) {
    for (final each in other) {
      add(each);
    }
  }

  List<T> concat<T>() {
    final List<T> lst = [];
    for (final each in this) {
      lst.addAll(each);
    }
    return lst;
  }

  List<T> filter<T>(bool Function(T elem) pred) {
    final List<T> lst = [];
    for (final each in this) {
      if (pred(each)) {
        lst.add(each);
      }
    }
    return lst;
  }

  int count() {
    return length;
  }

  List<T> myMap<T>(T Function(T elem) f) {
    final List<T> lst = [];
    for (final each in this) {
      lst.add(f(each));
    }
    return lst;
  }

  U foldl<T, U>(U initial, U Function(U acc, T elem) f) {
    for (final each in this) {
      initial = f(initial, each);
    }
    return initial;
  }

  U foldr<T, U>(U initial, U Function(T elem, U acc) f) {
    for (final each in reversed) {
      initial = f(each, initial);
    }
    return initial;
  }

  List<T> reverse<T>() {
    final List<T> lst = [];
    for (final each in reversed) {
      lst.add(each);
    }
    return lst;
  }
}
